packages/aws-cdk/lib/api/garbage-collection/garbage-collector.ts

```diff
@@ -0,0 +1,389 @@
+   */
+  readonly action: 'print' | 'tag' | 'delete-tagged' | 'full';
+
+  /**
+   * The type of asset to garbage collect.
+   */
+  readonly type: 's3' | 'ecr' | 'all';
+
+  /**
+   * The days an asset must be in isolation before being actually deleted.
+   */
+  readonly rollbackBufferDays: number;
+
+  /**
+   * The environment to deploy this stack in
+   *
+   * The environment on the stack artifact may be unresolved, this one
+   * must be resolved.
+   */
+  readonly resolvedEnvironment: cxapi.Environment;
+
+  /**
+   * SDK provider (seeded with default credentials)
+   *
+   * Will be used to make SDK calls to CloudFormation, S3, and ECR.
+   */
+  readonly sdkProvider: SdkProvider;
+
+  /**
+   * The name of the bootstrap stack to look for.
+   *
+   * @default DEFAULT_TOOLKIT_STACK_NAME
+   */
+  readonly bootstrapStackName?: string;
+
+  /**
+   * Max wait time for retries in milliseconds (for testing purposes).
+   *
+   * @default 60000
+   */
+  readonly maxWaitTime?: number;
+}
+
+/**
+ * A class to facilitate Garbage Collection of S3 and ECR assets
+ */
+export class GarbageCollector {
+  private garbageCollectS3Assets: boolean;
+  private garbageCollectEcrAssets: boolean;
+  private permissionToDelete: boolean;
+  private permissionToTag: boolean;
+  private bootstrapStackName: string;
+  private maxWaitTime: number;
+
+  public constructor(readonly props: GarbageCollectorProps) {
+    this.garbageCollectS3Assets = ['s3', 'all'].includes(props.type);
+    this.garbageCollectEcrAssets = ['ecr', 'all'].includes(props.type);
+
+    debug(`${this.garbageCollectS3Assets} ${this.garbageCollectEcrAssets}`);
+
+    this.permissionToDelete = ['delete-tagged', 'full'].includes(props.action);
+    this.permissionToTag = ['tag', 'full'].includes(props.action);
+    this.maxWaitTime = props.maxWaitTime ?? 60000;
+
+    this.bootstrapStackName = props.bootstrapStackName ?? DEFAULT_TOOLKIT_STACK_NAME;
+
+    // TODO: ECR garbage collection
+    if (this.garbageCollectEcrAssets) {
+      throw new Error('ECR garbage collection is not yet supported');
+    }
+  }
+
+  /**
+   * Perform garbage collection on the resolved environment.
+   */
+  public async garbageCollect() {
+    // SDKs
+    const sdk = (await this.props.sdkProvider.forEnvironment(this.props.resolvedEnvironment, Mode.ForWriting)).sdk;
+    const cfn = sdk.cloudFormation();
+    const s3 = sdk.s3();
+
+    const qualifier = await this.bootstrapQualifier(sdk, this.bootstrapStackName);
+    const activeAssets = new ActiveAssetCache();
+
+    // Grab stack templates first
+    const startTime = Date.now();
+    await refreshStacks(cfn, activeAssets, this.maxWaitTime, qualifier);
+    // Start the background refresh
+    const backgroundStackRefresh = new BackgroundStackRefresh({
+      cfn,
+      activeAssets,
+      qualifier,
+      maxWaitTime: this.maxWaitTime,
+    });
+    const timeout = setTimeout(backgroundStackRefresh.start, Math.max(startTime + 300_000 - Date.now(), 0));
```

Comment:
What I was thinking is that you would call:

```ts
backgroundRefresh.start();
```

And that it would manage its own timer.

By doing this, you're running into the same issue again that `setTimeout` returns a timer handle, and the first timer handle goes into the variable `timeout` here, and timer handles 2..N go into `this.timeout` on the class.

But instead, we would like for *all* timer handles to go into the same variable.

Reply:
ok i did that. the reason i had it this way was because i didn't want to start the refresh immediately after the first invocation of refreshStacks, but I think I did something reasonable to mitigate that now.